File: demo/demo-macos/demo-macos/PlayerViewController.m
//
//  PlayerViewController.m
//  demo-macos
//
//  Created by Single on 2017/3/15.
//  Copyright © 2017年 single. All rights reserved.
//

#import "PlayerViewController.h"
#import <SGPlayer/SGPlayer.h>

@interface PlayerViewController () <SGPlayerDelegate>

@property (nonatomic, strong) SGPlayer * player;

@property (weak) IBOutlet NSTextField *totalTimeLabel;
@property (weak) IBOutlet NSTextField *currentTimeLabel;
@property (weak) IBOutlet NSSlider *progressSilder;
@property (weak) IBOutlet NSButton *playButton;
@property (weak) IBOutlet NSButton *pauseButton;
@property (weak) IBOutlet NSTextField *stateLabel;
@property (weak) IBOutlet NSView *controlView;

@end

@implementation PlayerViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.view.wantsLayer = YES;
    self.view.layer.backgroundColor = [NSColor blackColor].CGColor;
    self.controlView.wantsLayer = YES;
    self.controlView.layer.backgroundColor = [NSColor colorWithWhite:0 alpha:0.5].CGColor;
    self.progressSilder.trackFillColor = [NSColor yellowColor];
}

- (void)setup
{
    NSURL * URL = [NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:@"i-see-fire" ofType:@"mp4"]];
    
    self.player = [[SGPlayer alloc] init];
    self.player.delegate = self;
    self.player.videoRenderer.view = self.view;
//    self.player.videoRenderer.displayMode = SGDisplayModeVR;
//    [self.player.videoRenderer setFrameOutput:^(SGVideoFrame * frame) {
//        NSLog(@"frame output : %f", CMTimeGetSeconds(frame.timeStamp));
//    }];
    [self.player replaceWithURL:URL];
    [self.player waitUntilReady];
    [self.player play];
}

- (IBAction)play:(id)sender
{
    [self.player play];
}

- (IBAction)pause:(id)sender
{
    [self.player pause];
}

- (void)player:(SGPlayer *)player didChangeStatus:(SGPlayerStatus)status
{
//    NSLog(@"%s, %ld", __func__, status);
}

- (void)player:(SGPlayer *)player didChangePlaybackState:(SGPlaybackState)state
{
//    NSLog(@"%s, playing  : %d, %d, %d", __func__, state & SGPlaybackStatePlaying, state & SGPlaybackStateSeeking, state & SGPlaybackStateFinished);
    if (state & SGPlaybackStateFinished) {
        self.stateLabel.stringValue = @"Finished";
    } else if (state & SGPlaybackStatePlaying) {
        self.stateLabel.stringValue = @"Playing";
    } else {
        self.stateLabel.stringValue = @"Paused";
    }
}

- (void)player:(SGPlayer *)player didChangeCurrentTime:(CMTime)currentTime duration:(CMTime)duration
{
//    NSLog(@"%s, %f", __func__, CMTimeGetSeconds(currentTime));
    self.progressSilder.doubleValue = CMTimeGetSeconds(currentTime) / CMTimeGetSeconds(duration);
    self.currentTimeLabel.stringValue = [self timeStringFromSeconds:CMTimeGetSeconds(currentTime)];
    self.totalTimeLabel.stringValue = [self timeStringFromSeconds:CMTimeGetSeconds(duration)];
}

- (void)player:(SGPlayer *)player didChangeLoadedTime:(CMTime)loadedTime loadedDuuration:(CMTime)loadedDuuration
{
//    NSLog(@"%s, %f, %f", __func__, CMTimeGetSeconds(loadedTime), CMTimeGetSeconds(loadedDuuration));
}

- (NSString *)timeStringFromSeconds:(CGFloat)seconds
{
    return [NSString stringWithFormat:@"%ld:%.2ld", (long)seconds / 60, (long)seconds % 60];
}

@end

File: demo/demo-ios/demo-ios-swift/PlayerViewController.swift
//
//  PlayerViewController.swift
//  demo-ios-swift
//
//  Created by Maksim Alyabyshev on 06.02.2018.
//  Copyright © 2018 single. All rights reserved.
//

import SGAVPlayer

class PlayerViewController: UIViewController {
    
    var player = SGAVPlayer.init()
    
    @IBOutlet weak var stateLabel: UILabel!
    @IBOutlet weak var progressSilder: UISlider!
    @IBOutlet weak var currentTimeLabel: UILabel!
    @IBOutlet weak var totalTimeLabel: UILabel!
    var isProgressSilderTouching = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.black
        sg_registerNotification(
            for: player,
            playbackStateAction: #selector(playbackStateAction(_:)),
            loadStateAction: #selector(loadStateAction(_:)),
            currentTimeAction: #selector(currentTimeAction(_:)),
            loadedAction: #selector(loadedTimeAction(_:)),
            errorAction: #selector(errorAction(_:))
        )
        view.insertSubview(player.view, at: 0)
        let contentURL = URL(fileURLWithPath: Bundle.main.path(forResource: "i-see-fire", ofType: "mp4") ?? "")
        //    NSURL * contentURL = [NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:@"google-help-vr" ofType:@"mp4"]];
        player.replace(withContentURL: contentURL)
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        player.view.frame = view.bounds
    }
    
    @IBAction func back(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    @IBAction func play(_ sender: Any) {
        player.play()
    }
    @IBAction func pause(_ sender: Any) {
        player.pause()
    }
    @IBAction func progressTouchDown(_ sender: Any) {
        isProgressSilderTouching = true
    }
    @IBAction func progressTouchUp(_ sender: Any) {
        isProgressSilderTouching = false
        player.seek(toTime: TimeInterval(Float(player.duration) * progressSilder.value))
    }
    
//    /**
//     *  Notification Name
//     */
//    SGPLAYER_EXTERN NSString * const SGPlayerPlaybackStateDidChangeNotificationName;
//    SGPLAYER_EXTERN NSString * const SGPlayerLoadStateDidChangeNotificationName;
//    SGPLAYER_EXTERN NSString * const SGPlayerCurrentTimeDidChangeNotificationName;
//    SGPLAYER_EXTERN NSString * const SGPlayerLoadedTimeDidChangeNotificationName;
//    SGPLAYER_EXTERN NSString * const SGPlayerDidErrorNotificationName;
//    /**
//     *  Notification Userinfo Key
//     */
//    SGPLAYER_EXTERN NSString * const SGPlayerNotificationUserInfoObjectKey;    // Common Object Key.

    @objc func playbackStateAction(_ notification: NSNotification) {
        let userInfo: [AnyHashable : Any]? = notification.userInfo!
        let playbackStateModel = userInfo!["SGPlayerNotificationUserInfoObjectKey"] as! SGPlaybackStateModel
        print("====> STATE  playback  SGPlaybackStateModel:", playbackStateModel.current ," = ", playbackStateModel.current.rawValue)
        
        var text: String
        switch playbackStateModel.current {
        case SGPlayerPlaybackState.idle:
            text = "Idle"
        case SGPlayerPlaybackState.playing:
            text = "Playing"
        case SGPlayerPlaybackState.seeking:
            text = "Seeking"
        case SGPlayerPlaybackState.paused:
            text = "Paused"
        case SGPlayerPlaybackState.interrupted:
            text = "Interrupted"
        case SGPlayerPlaybackState.stopped:
            text = "Stopped"
        case SGPlayerPlaybackState.finished:
            text = "Finished"
        case SGPlayerPlaybackState.failed:
            text = "Failed"
        }
        stateLabel.text = text
    }
    
    @objc func loadStateAction(_ notification: Notification) {
        let userInfo: [AnyHashable : Any]? = notification.userInfo!
        let loadStateModel = userInfo!["SGPlayerNotificationUserInfoObjectKey"] as! SGLoadedStateModel
        print("====> STATE  loaded  SGLoadedStateModel:", loadStateModel.current ," = ", loadStateModel.current.rawValue)

        if loadStateModel.current.rawValue == SGPlayerLoadState.playable.rawValue {
            if player.playbackState.rawValue == SGPlayerLoadState.idle.rawValue {
                player.play()
            }
        }
    }
    
    @objc func currentTimeAction(_ notification: Notification) {
//        let currentTimeModel: SGTimeModel? = notification.userInfo.sg_currentTimeModel()
        let userInfo: [AnyHashable : Any]? = notification.userInfo!
        let currentTimeModel = userInfo!["SGPlayerNotificationUserInfoObjectKey"] as! SGTimeModel
        print("====> STATE  time  currentTimeAction:", currentTimeModel.current ," = ", currentTimeModel.current.hashValue)
        
        if !isProgressSilderTouching {
            progressSilder.value = Float(currentTimeModel.percent)
        }
        currentTimeLabel.text = timeString(fromSeconds: CGFloat(currentTimeModel.current))
        totalTimeLabel.text = timeString(fromSeconds: CGFloat(currentTimeModel.duration))
        print("====> STATE  time  totalTimeLabel:", currentTimeModel.duration)
    }
    @objc func loadedTimeAction(_ notification: Notification) {
//        let loadedTimeModel: SGTimeModel? = notification.userInfo?.sg_loadedTimeModel()
        let userInfo: [AnyHashable : Any]? = notification.userInfo!
        let loadedTimeModel = userInfo!["SGPlayerNotificationUserInfoObjectKey"] as! SGTimeModel
        print("====> STATE  time  loadedTimeAction:", loadedTimeModel.current ," = ", loadedTimeModel.current.hashValue)
    }
    @objc func errorAction(_ notification: Notification) {
//        let error: Error? = notification.userInfo?.sg_error()
        let error = player.error
        print("====> ERROR: \(String(describing: error))")
    }
    func timeString(fromSeconds seconds: CGFloat) -> String {
        let minutes = String(Int((seconds / 60)))
        var seconds = String(Int(seconds) % 60)
        if seconds.count < 2 {
            seconds.insert("0", at: seconds.startIndex)
        }
        return minutes+":"+seconds
    }
    
    deinit {
        sg_removeNotification(for: player)
    }

}

